qcodes/instrument_drivers/tektronix/Keithley_s46.py

```diff
@@ -0,0 +1,225 @@
+        """
+        if new_state == "close":
+            self._relay_lock.acquire(self._channel_number)
+        elif new_state == "open":
+            self._relay_lock.release(self._channel_number)
+
+        self.write(f":{new_state} (@{self._channel_number})")
+
+    @property
+    def channel_number(self) -> int:
+        return self._channel_number
+
+
+class S46(VisaInstrument):
+
+    relay_names: list = ["A", "B", "C", "D"] + [f"R{j}" for j in range(1, 9)]
+
+    # Make a dictionary where keys are channel aliases (e.g. 'A1', 'B3', etc)
+    # and values are corresponding channel numbers.
+    aliases: Dict[str, int] = {
+        f"{a}{b}": count + 1 for count, (a, b) in enumerate(product(
+            ["A", "B", "C", "D"],
+            range(1, 7)
+        ))
+    }
+    aliases.update({f"R{i}": i + 24 for i in range(1, 9)})
+
+    def __init__(
+            self,
+            name: str,
+            address: str,
+            **kwargs: Any
+    ):
+
+        super().__init__(name, address, terminator="\n", **kwargs)
+
+        channels = ChannelList(
+            self,
+            "channel",
+            S46Channel,
+            snapshotable=False
+        )
+
+        for relay_name, channel_count in zip(
+                S46.relay_names, self.relay_layout):
+
+            relay_lock = RelayLock(relay_name)
+
+            for channel_index in range(1, channel_count + 1):
+                # E.g. For channel 'B2', channel_index is 2
+                if channel_count > 1:
+                    alias = f"{relay_name}{channel_index}"
+                else:
+                    alias = relay_name  # For channels R1 to R8, we have one
+                    # channel per relay. Channel alias = relay name
+
+                channel_number = S46.aliases[alias]
+                channel = S46Channel(self, alias, channel_number, relay_lock)
+                channels.append(channel)
+                self.add_submodule(alias, channel)
+
+        self.add_submodule("channels", channels)
+        self.connect_message()
+
+    @cached_method
+    def get_closed_channel_numbers(self) -> List[int]:
+        """
+        Query the instrument for closed channels. Add an option to return
+        a cached response so we can prevent this method from being called
+        repeatedly for each channel during initialization of the driver.
+        """
+        closed_channels_str = re.findall(r"\d+", self.ask(":CLOS?"))
+        return [int(i) for i in closed_channels_str]
+
+    def get_closed_channels(self) -> List[S46Channel]:
+        """
+        Return a list of closed channels as a list of channel objects
+        """
+        return [
+            channel for channel in self.channels if
+            channel.channel_number in self.get_closed_channel_numbers()
+        ]
+
+    def open_all_channels(self) -> None:
+        """
+        Please do not write ':OPEN ALL' to the instrument as this will
+        circumvent the lock.
+        """
```

Comment:
Is this slow compared to `:OPEN ALL`? Is that an issue or perhaps it doesn't matter?

Reply:
Its probably slower, but I do not think it matters.